packages/forms/src/UIForm/UIForm.container.js

```diff
@@ -5,28 +5,94 @@ import React from 'react';
 import UIFormTranslatedComponent from './UIForm.component';
 import { formPropTypes } from './utils/propTypes';
 
+/**
+ * add error object on a formSchema if it doesn't exist
+ * @param {FormSchema} formSchema
+ * @return {FormSchema}
+ */
+function addErrorObject(formSchema) {
+	if (!formSchema.errors) {
+		return { errors: {}, ...formSchema };
+	}
+	return formSchema;
+}
+
+/**
+ * reinit liveState with empty errors if it doesn't exist
+ * @param {FormSchema} newFormSchema
+ * @param {State} prevState
+ * @return {State}
+ */
+const reinitLiveState = newFormSchema => () => ({
+	liveState: addErrorObject(newFormSchema),
+});
+
+/**
+ * update live state with new properties derived from user interacting
+ * with the form
+ * @param {Object} properties
+ * @param {State} prevState
+ * @return {State}
+ */
+const setLiveStateProperties = properties => prevState => ({
+	...prevState,
+	liveState: { ...prevState.liveState, properties },
+});
+
+/**
+ * update liveState formSchema with errors
+ * @param {Object} errors
+ * @param {State} prevState
+ * @return {State}
+ */
+const setLiveStateErrors = errors => prevState => ({
+	...prevState,
+	liveState: { ...prevState.liveState, errors },
+});
+
+/**
+ * update initialState with liveState after a user submission of the form
+ * @param {Object} newProperties
+ * @param {State} prevState
+ * @return {State}
+ */
+const setLiveAsInitialState = prevState => ({
+	...prevState,
+	initialState: prevState.liveState,
+});
+
+/**
+ * update liveState with initialState, reseting form
+ * @param {State} prevState
+ * @return {State}
+ */
+const setInitialStateAsLiveState = prevState => ({
+	...prevState,
+	liveState: prevState.initialState,
+});
+
 export default class UIForm extends React.Component {
 	static displayName = 'Container(UIForm)';
 	constructor(props) {
 		super(props);
-		this.state = { ...this.props.data };
-		if (!this.state.errors) {
-			this.state.errors = {};
-		}
+		this.state = {
+			initialState: addErrorObject(this.props.data),
```

Comment:
interesting solution to separate the initial data with the living data. So, addErrorObject make a spread on the first level. We lose the reference on the first level but if the formdata has nested data, we have a reference with initialState, Would  it is interesting to make a deep shallow to avoid that ?

Reply:
using https://lodash.com/docs/4.17.11#cloneDeep for example ?

Reply:
yes

Reply:
But everything is treated as immutable. If the user get it and mutate data, it's a bad practice and he has to fix that.

Reply:
im not sure that we need a deep clone here and on each state changes:
- because it wasn't done before and it worked fine
- because PureComponent and other méthod usualy do a shallow check on props